### ISSUE TITLE
model "paths" with steady forcing depend on start year 

### ISSUE THREAD

**Issue by tkleiner** [pism/pism#407]
I run the `circular_withshelf.nc`set-up again to confirm if model results after the `-bed_def lc` update in stable0.7.3 from stable1.0 are reproducible. I found that all my 10 runs (serial, no restarts) started at -325 ka (yellow to light green or left column in the legend of the figure) followed the same path in e.g. `slvol`. At that time a was very happy and thought I could reduce the runtime by starting at -220 ka. Again all of my 10 runs followed the same path (middle column of the legend). Unfortunately the path of the 325er is different to the path of the 220er. I run this experiment also for 135ka (right column of the legend) and again the paths are different to the 325er and 220er. 

I decided to run the 135er again on different nodes and different days and all 30 of these runs followed the same path (not shown). 

I do not understand what is causing these differences. Any help is very welcome. 

Pism version: `:source = "PISM stable v0.7.3-20-g9cfa7cb"`

Relevant part of the slurm batch file:
```
ys=-135000
#ys=-220000
#ys=-325000

name=bed_lc_fftw_estimate_${SLURM_JOB_ID}

if [ ! -e circular_withshelf.nc ] ; then
  $PISM_SRC_DIR/examples/marine/circular/circular_ice_sheet.py
fi

srun  \
     pismr -options_left -verbose 2 \
     -i circular_withshelf.nc -bootstrap -mx 61 -my 61 -lz 5500 -lbz 2000 -mz 41 -mbz 11  \
     -sia_e 2.0 -atmosphere given -surface given -ocean pik -meltfactor_pik 5e-3 \
     -bed_def lc -ssa_method fd -ssa_e 0.55 -pik -calving eigen_calving,thickness_calving \
     -eigen_calving_k 1.0e17 -thickness_calving_threshold 150.0 -stress_balance ssa+sia \
     -pseudo_plastic -pseudo_plastic_q 0.6 -till_effective_fraction_overburden 0.02 \
     -tauc_slippery_grounding_lines -topg_to_phi 10.0,30.0,-700.0,200.0 -hydrology null \
     -ys $ys -ye 0 \
     -ts_file ${name}_ts.nc -ts_times yearly \
     -extra_file ${name}_ex.nc -extra_times 1000 \
     -extra_vars thk,usurf,mask,bmelt,tillwat,temppabase,tempicethk_basal,csurf,cbase,topg \
     -o ${name}.nc -o_size big
``` 
![slvol-fftw_estimate_1core](https://user-images.githubusercontent.com/5449934/39531119-c6236940-4e2a-11e8-8dfd-6dddec35c2bc.png)

**Comment by talbrecht** [comment]
Maybe this issue is connected to issue post #370, in which intermediate restarts alter the final simulation result?! Can we narrow down the source of this issue? Does this problem also occur in experiments without -pik option or -bed_def lc?

**Comment by talbrecht** [comment]
I spent some more time on testing for possible causes of this start-year-issue. I can say it has nothing to do with the choice of `-bed_def`, nor with `-pik` options or the `-ocean`. In other words, `-ocean pik -meltfactor_pik 0 -bed_def none` still shows the start-year dependency in the time series. When using `-yield_stress constant` we can get rid of this issue, so now we know where to look at.

I found that the combination of `-tauc_slippery_grounding_lines` and the feedback loop of enthalpy-derived `bmelt` and basal resistance model via the `-hydrolgy null` model creates some asymmetry, which amplifies over time if `tillwat` does not vanish. Somehow, this instability is dependent on start time.

Based on the batch run script by @tkleiner above, if you replace `-tauc_slippery_grounding_lines` by `-hydrology_use_const_bmelt -hydrology_const_bmelt 0` one gets x-y-symmetrical and reproducible results independent of the start year. However, if `bmelt` is chosen large enough (here default 1cm/yr: `-hydrology_const_bmelt 3.168876461e-10`) with respect to the `hydrology_tillwat_decay_rate`, then `tillwat`>0 can accumulate, and symmetry still breaks, which means that there is still some feedback active...

**Comment by nihaoyanghu** [comment]
I have run 200 years Greenland simulation with different restart period. It came out that every simulation is different if the restart point is different. 

I also tried simulations which do not apply -bed_def, the similar problem still exist.  I checked the very first year's result when the model restart, and compare it with simulation without a restart. It seems that the snow_depth have the largest difference even though other variables like tillwat, shelfbtemp, thk, bheatflx also have differences. 
![snow_depth_dif](https://user-images.githubusercontent.com/30825864/46607047-c677a080-caff-11e8-9d5d-ec6ea404e3c9.jpeg)

The upper figure give the snow_depth different between  the end of the first year when the model restart and the normal simulation without restart. This is strange.

**Comment by tkleiner** [comment]
Thank you for reporting this. Could you please provide the pism version (e.g. via `pismr | grep PISMR`) and a simple run script (or the pism command) to reproduce this behavior?

**Comment by nihaoyanghu** [comment]
[hyang@ollie1:~/work/PISM/Hu/glacial_index2/test3]$ pismr | grep PISMR
PISMR stable v0.7.3-20-g9cfa7cb (basic evolution run mode)

You can find these two run at

/home/ollie/hyang/work/PISM/Hu/glacial_index2/test3

This is the run with 20 years restart 

srun -n 144 pismr -i /home/ollie/hyang/work/PISM/Hu/input/Greenland_5km_v1.1.nc -bootstrap -Mx 300 -My 560 -Mz 201 -Mbz 21 -Lz 4000 -Lbz 2000 -skip -skip_max 10 -ssa_rtol 1e-4 -ssafd_ksp_rtol 1e-5 -calendar 365_day -config /home/ollie/hyang/software/pism-intel18-intelmpi/awi0.7-dev/lib/pism/pism_config.nc -stress_balance ssa+sia -sia_e 5 -pseudo_plastic -till_effective_fraction_overburden 0.01 -hydrology_tillwat_max 1 -pik -calving eigen_calving,thickness_calving -thickness_calving_threshold 200.0 -surface pdd -atmosphere index -atmosphere_index_climate_file /home/ollie/hyang/work/project/AntarcticForcing/Nlu/climate_forcing_Greenland_5km_127k-21k.nc -atmosphere_index_index_file /home/ollie/hyang/work/project/AntarcticForcing/Nlu/glac_index_NGRIP_GLsyn_127k-21k.nc -temp_lapse_rate 5 -ocean pik,delta_SL -meltfactor_pik 1e-2 -ocean_delta_SL_file /home/ollie/lniu/data/LongIndex/RSL_Rohling2014_5.3Ma_PISM.nc -ys -127000 -ye -126980 -extra_file ../output/ex_Greenland.nc -extra_times -127000:1:-126980 -extra_vars thk,usurf,mask,air_temp,precipitation -ts_file ../output/ts_Greenland.nc -ts_times -127000:1:-126980 -o ../output/out_Greenland_-126980.nc -o_size small -save_file ../output/snap -save_times -127000:1:-126980 -save_size small -save_split

This is the run with 100 year restart

srun -n 144 pismr -i /home/ollie/hyang/work/PISM/Hu/input/Greenland_5km_v1.1.nc -bootstrap -Mx 300 -My 560 -Mz 201 -Mbz 21 -Lz 4000 -Lbz 2000 -skip -skip_max 10 -ssa_rtol 1e-4 -ssafd_ksp_rtol 1e-5 -calendar 365_day -config /home/ollie/hyang/software/pism-intel18-intelmpi/awi0.7-dev/lib/pism/pism_config.nc -stress_balance ssa+sia -sia_e 5 -pseudo_plastic -till_effective_fraction_overburden 0.01 -hydrology_tillwat_max 1 -pik -calving eigen_calving,thickness_calving -thickness_calving_threshold 200.0 -surface pdd -atmosphere index -atmosphere_index_climate_file /home/ollie/hyang/work/project/AntarcticForcing/Nlu/climate_forcing_Greenland_5km_127k-21k.nc -atmosphere_index_index_file /home/ollie/hyang/work/project/AntarcticForcing/Nlu/glac_index_NGRIP_GLsyn_127k-21k.nc -temp_lapse_rate 5 -ocean pik,delta_SL -meltfactor_pik 1e-2 -ocean_delta_SL_file /home/ollie/lniu/data/LongIndex/RSL_Rohling2014_5.3Ma_PISM.nc -ys -127000 -ye -126900 -extra_file ../output/ex_Greenland.nc -extra_times -127000:1:-126900 -extra_vars thk,usurf,mask,air_temp,precipitation -ts_file ../output/ts_Greenland.nc -ts_times -127000:1:-126900 -o ../output/out_Greenland_-126900.nc -o_size small -save_file ../output/snap -save_times -127000:1:-126900 -save_size small -save_split

**Comment by talbrecht** [comment]
This is an example with reduced options (no` -pik`, no `-bed_def` etc.), which still shows the startyear-dependency (as loop on the same cpu!) for PISM versions `stable v0.7.3-20-g9cfa7cb30` and  `stable v1.0-35-g9b1b6ff1c`, but also for latest dev  `development v1.0-729-g761f01898` :

```
for ys in 0 100000 200000; do
     length=100000
     name=startyearbug_boot_${ys}
     srun  \
     $PISM_SRC_DIR/pismr -options_left -verbose 2 \
     -i circular_withshelf.nc -bootstrap -mx 61 -my 61 -lz 6000 -lbz 2000 -mz 41 -mbz 11 \
     -surface given \
     -bed_def none -hydrology null -calendar none \
     -ssa_method fd -stress_balance ssa+sia \
     -calving ocean_kill -ocean_kill_file circular_withshelf.nc \
     -ys $((ys)) -y $((length)) \
     -ts_file ${name}_ts.nc -ts_times yearly \
     -extra_file ${name}_ex.nc -extra_times 500 \
     -save_file ${name}_save -save_times 5000 -save_split -save_size medium \
     -extra_vars thk,usurf,mask,bmelt,tillwat,temppabase,tempicethk_basal,topg,tauc,tillphi \
     -o ${name}.nc -o_size big >> out_$ys.out
done
```
The first save_file is identical using` util/nccmp.py -x -v time,timestamp s1_yr0.nc s2_yr0.nc`, but differences show up during runtime. Hence, it seems, that it is not due to the `-bootstrap` procedure, as `-i` from the same PISM-readable file shows the same problem. 

I don't know if this helps, but why and where is part_grid set to true in this example? I tried to use option `-no_part_grid` or `-config_override`? I just wonder, because `Href / ice_area_specific_volume` is among the diverging fields, even though` -pik` option is not used here!

![repro_dev_1cpu](https://user-images.githubusercontent.com/1190987/47187616-30c8e600-d335-11e8-8f37-b76895b71cd3.png)

**Comment by tkleiner** [comment]
Torsten, 
which PISM version and variable has been used to create the upper panel (sea-level ...) of your figure? If I run PISM stable v0.7.3-20-g9cfa7cb and plot "slvol" the figures is completely different. Ice volume in m s.l.e (slvol) is increasing at the beginning.

**Comment by talbrecht** [comment]
Thomas,
the plot above was `slvol` with the PISM stable v0.7.3-20-g9cfa7cb for your code example (as in the first comment of this issue from May 2nd). For my modified run script (October 19th) ice volume is actually increasing up to 53.4 m SLE after 7800 model years and then stabilized at around 44 m SLE. But the point of my plot here was the difference between different start years, also for disabled bed deformation.

**Comment by tkleiner** [comment]
Thank you for clarification. Your figure is based on my old set-up. 

I rerun your "reduced options"   experiment (https://github.com/pism/pism/issues/407#issuecomment-431185830) with `-lz 6500` to avoid yesterdays max thickness error. I did this three times now and found that in the figure all lines from the slvol output overlay each other. Thus, the start year dependency is small compared to the runs with `-pik -bed_def lc`. Nevertheless, there are differences in ice thickness between runs with different start years. 

startyearbug_boot_0.nc versus:
- startyearbug_boot_-200000.nc: Variable thk: delta = 4.256791e+00
- startyearbug_boot_-100000.nc: Variable thk: delta = 4.297947e+00
- startyearbug_boot_100000.nc: Variable thk: delta = 4.297947e+00
- startyearbug_boot_200000.nc: Variable thk: delta = 4.256791e+00

Again, -100kyrs/100kyrs and -200kyrs/200kyrs show the same behaviour.
Files with the same start year but from different runs are identical (reproducible). 

I do not understand, why yesterdays (https://github.com/pism/pism/issues/407#issuecomment-435912500) runs start to deviate if the ice thickness is close to the allowed maximum.

**Comment by tkleiner** [comment]
> I don't know if this helps, but why and where is part_grid set to true in this example? I tried to use option `-no_part_grid` or `-config_override`? I just wonder, because `Href / ice_area_specific_volume` is among the diverging fields, even though` -pik` option is not used here!

part_grid and kill_icebergs are set to true in `src/base/util/PISMConfigInterface.cc`, because  `-calving ocean_kill` is used in your example.

**Comment by ckhroulev** [comment]
I'm not sure if this really *is* the issue, but it is worth pointing out that the circular setup used here is very sensitive to the computed ice velocity. Any asymmetry in ice velocity will result in an asymmetry in ice geometry, which will affect its velocity computed during the next time step, etc.

PISM v0.7 included ice shelves in the load used by bed deformation models and the response of the Lingle-Clark model (unlike point-wise isostasy) is non-local, so asymmetries in shelves result in larger asymmetries in the bed topography. This can move the grounding line and so affect the shelf, escalating this process.

The same logic applies when `-tauc_slippery_grounding_lines` is set: changes in shelf velocity and geometry affects grounded ice which then affects the shelf, etc.

You can try setting `-ssafd_ksp_rtol` to a small number,  e.g. `1e-12`. This should make the SSA code try harder, making the ice velocity field more symmetric and better-behaved.

How is this related to the start time? Probably through the code choosing the time step length.

So, I think it would be nice to find a test case less sensitive to tiny perturbations of the model state.

Finally, if you want to get to the bottom of this we need a test case that takes roughly 30 seconds to run on a 4-core CPU. Test cases that involve 100,000-year runs are very hard to use.

PS: Sorry it took me so long to respond -- I'm working on PISM half-time these days (the other half is spent taking care of our 6 month old baby).

**Comment by talbrecht** [comment]
Regarding the test case, Thomas pointed out, that differences in time stepping occur already after the first time step, so a few model years may be sufficient for a PISM test. And I have run stable version 1.1 and still find time steps depending on the choice of the start year (tiny difference of course 10 orders of magnitude smaller than the timestep itself). The question is now, do those differences occur only in the reporting (timeseries) at different time steps or is the model result as such a different one? 

You are right, that this circular setup seems to be a quite sensitive, but I would say that the high sensitivity helps us to understand the limits of reproducibility, and the implications of small differences in a highly nonlinear system after many iterations.

PS: Constantine, congratulations again.. I'm happy for you and your family.

**Comment by talbrecht** [comment]
My hope in this fix is, that it could also solve a slightly different reproducibility issue as pointed out in the end of #370 (following the [run scripts](https://github.com/pism/pism/issues/370#issuecomment-373191020) by @tkleiner). Many users run simulations in comparably short classes which require restarts in between. The solution should be the same as a continuous run from start to endyear. The (sensitive) circular setup example shows how solutions can diverge in this two cases (continuously for 20,000 years in grey vs. 20 restarts for 100 years each in colors), also for stable version 1.1 (here calculated with` -bed_def lc` and 8 CPU for 30km resolution):

![restart_issue_stable1p1](https://user-images.githubusercontent.com/1190987/50078553-93304c80-01e7-11e9-8250-f5f7adaf1059.png)

**Comment by talbrecht** [comment]
A brief update in the divergence between "one-go" vs. "20-times-restart" simulations: Using the temporary fix from #422 for the 356_day calendar shows the same result. 
What happens in the "one-go" simulation is, that the circular ice sheet retreats first causing the bed uplift by a few hundred meters and ice shelf formation. When ice shelves reach the boundary of the computational domain (this is of course not realistic) they can add buttressing to the ice sheet flow and hence a regrowth. In the restart case, bed uplift responds slightly slower and somehow ice shelves do not extend as much as in the the "one-go" experiment. In this particular example, this amplifies the small differences. 

I then switched off bed deformation using -`bed_def none`, and found almost no divergence in the solution. My guess is that in this case threshold behavior (amplifying small differences) does not show up in the simulations and hence no (or only tiny) divergence of simulation results.

![restart_issue_stable1p1_nolc](https://user-images.githubusercontent.com/1190987/50212860-66616e00-037c-11e9-9fa6-b17f0f2c809c.png)

**Comment by ckhroulev** [comment]
@talbrecht Thanks for the updates!

Note, though, that tests running PISM for 20000 years are not useful for me. (I know I'm repeating myself here.)

If you would like to help with this, please create a test that demonstrates the issue and uses short runs (if that's possible -- I know it isn't always easy). If reducing the run length is hard, consider reducing grid resolution to speed things up.

**Comment by talbrecht** [comment]
OK, in a coarse 60km resolution case a 10x100 year-simulation requires for 1CPU about 1-2 minutes: [bash-testcase.sh](https://github.com/pism/pism/files/2703198/exp-loc.txt)

**Comment by talbrecht** [comment]
We were having a computer science intern (@maxluebke), at PIK who has run some more tests regarding the startyear issue by building checksums at each timestep for different model variables and found a potential source of inconsistency in the adaptive time stepping scheme in `icemodel/timestepping.cc`. For extra files and result files output (not for intermediate savefiles) the timestep is adjusted close to the target time, such that for the exact target year a file can be written (hit multiples of a year).  
  
In `Time::step()` in `util/Time.cc` a small timestep `dT` is added to a comparably large time number `m_time_in_seconds`, both in units of seconds. For double precision this summation may imply an unavoidable rounding problem, as the time steps are calculated with decimal places, and the calculation of the current time step may hence differ with respect to the 'real' result and the floating point arithmetic, leading to different stored current times depending on the chosen start year. And as after each step, this time step is added to the current model time, small differences in time stepping can cumulate to comparably large differences in simulation results.  
  
We have tested the summation for different `m_time_in_seconds` and same `dT` in a small c-skript  [test_pism_dt.c](https://github.com/pism/pism/files/3276275/test_pism_dt.c.txt) (courtesy [Stefan Petri](https://www.pik-potsdam.de/members/petri/petri.html), remove .txt) and found for common optimization compiler flags same differences in summation results as in the PISM tests. In order to avoid this problem for commonly used optimization, we suggest to change time-keeping variables **from double to 64-bit integers**. It appears unlikely that PISM really needs fractions-of-second for its time keeping. This would affect in `util/Time.hh` the class members `Time::m_year_length` , `Time::m_time_in_seconds` , `Time::m_run_start` ,`Time::m_run_end` , and in `icemodel/IceModel.hh` the members `IceModel::m_dt` , `IceModel::t_TempAge IceModel` , `dt_TempAge`. Possibly also in `util/Timeseries.hh Timeseries::m_time` and `Timeseries::m_time_bounds,` also all the corresponding methods and functions.  

Using integer time steps would solve this issue, are there arguments to avoid this modifications? Let's discuss!


As an example, MOM5 or MOM6 make use of time management infrastructure with the integer datatype (in Fortran90)  
  
type time_type  
private  
integer:: seconds  
integer:: days  
integer:: ticks  
end type time_type  
  
with operators to add, compare and substract objects of this type, and functions to convert internal from/into external representations for different calenders (THIRTY_DAY_MONTHS, JULIAN, GREGORIAN, NOLEAP), see [https://github.com/mom-ocean/MOM5/blob/99168b44ab45f4f5b4fa2544a0c3f644f0afb666/src/shared/time_manager/time_manager.F90](https://github.com/mom-ocean/MOM5/blob/99168b44ab45f4f5b4fa2544a0c3f644f0afb666/src/shared/time_manager/time_manager.F90))

**Comment by sebhinck** [comment]
A colleague of me, Lennert, also just stumbled upon this bug. Both runs, which were starting from different start years, quickly diverged.

When investigating the problem I read this issue and the last comment of @talbrecht, suggesting that the problem comes from rounding errors when adding doubles, which will slowly accumulate and be amplified from other model components.

I had an idea on how to simply overcome this issue, without changing how the time accounting in PISM works. Lennert tested my modifications using his setup, and it seems to be working. 

I agree with @talbrecht that we do not need to care about fractions of seconds in an ice sheet model, and that therefore it would be sufficient to just add integers to keep track of the model time. However, this change would require some work on PISM's core. 

The "problem" is that the adaptive time stepping mechanism collects the maximum timestep as floating-point numbers for the different sub-models. This maximum timestep just reports the maximum value, but it can be smaller. Therefore we can simply round this value to the next lower whole second without breaking any stability criterion. The benefit is that each timestep hits a whole second, and therefore we never have to deal with any rounding problems.

Instead of changing all this in every sub-model, I applied the fix in the class MaxTimestep, which is used for handling the timestepping. My fix only changes 3 lines of code. 

I created this pull request: #458

I think that the change can not have any bad impact on other parts of the model. Or did I miss something? This is a quick and dirty fix to overcome this issue now, on the long-term, however, I would support the suggested solution of @talbrecht.

**Comment by talbrecht** [comment]
Thanks @sebhinck for suggesting this simple fix #458. I tested it in the dev branch with the above mentioned short [bash run script](https://github.com/pism/pism/files/2703198/exp-loc.txt). However, I could't find less divergence, even without bed deformation. Did it really solve the divergence in your more realistic setups or just a bit of it?

**Comment by sebhinck** [comment]
Hi @talbrecht , as far as I remember it solved the problem with the diverging paths. But I didn't try the example script, but Lennert tested my code for his setup.
Your example script, however, does test for reproducibility "restart vs one-go", right? While my fix is purely about the start-year issue, which Thomas initially reported. I did not test for the restarting issue.

**Comment by sebhinck** [comment]
I checked again, the same PISM setup starting once at year -32500 and once at year 1, was diverging using default PISM. 
When using #458 the results seem to be the same. `ncdiff` returns zeros for all diagnostics that I have tested.

**Comment by talbrecht** [comment]
I understand, the experiment was designed for a different issue, but the problem is likely the same, the rounding error with double precision. I ran another [simple experiment](https://github.com/pism/pism/files/5332461/exp-loc-startyear.txt) based on dev 9964da542 (just 1 minute on 1 CPU core), and actually found that your fix #458 works in that case. However, it fails in the "restart vs one-go" experiment. Could a similar `std::floor()` command in `src/util/Time.cc` help in that case?

**Comment by sebhinck** [comment]
You are talking about `Time::step(double delta_t)`? I don't think this would make a difference since MaxTimestep already took care of that, i.e. all Timesteps are already rounded, and, assuming the start date was a whole number, `m_time_in_seconds` should always be whole.
Could the restart issue be related to loss of precision when writing and reading data from disc?

**Comment by talbrecht** [comment]
Yes, you are right. My guess is, that for each of the result files (`-o`) small round-off errors occur (also for `-extra_file`, but nor for `-save_file`, as indicated above), which accumulate after some steps compared to the "one-go" case. Yes, writing and reading could be another source of error.

**Comment by sebhinck** [comment]
It seems that the only difference between reporting `save_file` and `extra_file` is the respective config flag: `time_stepping.hit_{save|extra}_times`, which is by default `yes` for extra files and `no` for save files (and also ts files).

What happens if we do not force the extra time to hit multiples?

Then the only difference might be that the restart run, obviously, forces the model to hit the end year of each chunk. But we could try to also force the "one-go" run to also hit this year by specifying `time_stepping.hit_multiples`.

For the restart run:
`-time_stepping.hit_extra_times no`

and for the one-go run:
`-time_stepping.hit_multiples X`, where `X` is the run length (years) of each chunk of the restart run.

Could you try this @talbrecht ? You have the script already running, and I only have an unstable internet connection right now...
I think this would bring both runs in perfect sync.

**Comment by talbrecht** [comment]
Yes, this works just fine, great. I added the modified [restart test script](https://github.com/pism/pism/files/5337233/exp-loc-restart.txt). Now we know fore sure...
However, in practical experience you may want to compare one long run with one that has been canceled and restarted at some arbitrary time step that you may not know before.

**Comment by sebhinck** [comment]
I think you got something wrong in your script. You are comparing two "one-go" runs, not "one-go vs restart".
I did this run [exp-loc-hit_multiples.txt](https://github.com/pism/pism/files/5339800/exp-loc-hit_multiples.txt), but have the feeling that the hit-multiple mechanism is not working as it should. I will open another issue for that..

**Comment by sebhinck** [comment]
Sorry, I deleted my previous message, because I found a typo in my runscript...

However, in your runscript, @talbrecht , you compare two "one-go" runs instead "one-go vs restart".
I did this comparison ([exp-loc-hit_multiples.txt](https://github.com/pism/pism/files/5340260/exp-loc-hit_multiples.txt)) but both results are still different, but the order of magnitude is very different:
with hit_multiples ~ 1e-11, without hit multiples: ~ 1e2

It is a bit disturbing that within 1000years the choice of the timestep has such a big impact on the experimental outcome.

**Comment by talbrecht** [comment]
Well, I wasn't looking at the ncdiff results but at the nccmp.py check2

**Comment by sebhinck** [comment]
Ok, here are the differences I got from `nccmp.py -v topg,thk   results/beddef_lc_60km_out9.nc     results/beddef_lc_60km_outf.nc`.

- "one-go" run *with* `-time_stepping.hit_multiples 100` : 
```
Variable topg: delta = 2.353318e-11, tol = 0.000000e+00
Variable thk: delta = 3.784749e-09, tol = 0.000000e+00
Files are different.
```

- "one-go" run *without* `-time_stepping.hit_multiples` :
```
Variable topg: delta = 6.879220e+00, tol = 0.000000e+00
Variable thk: delta = 7.247994e+02, tol = 0.000000e+00
Files are different.
```

**Comment by talbrecht** [comment]
Yes, I get similar small deltas (O(1e-9)) in the `-time_stepping.hit_multiples 100` case, which is a result of the LC bed deformation model, but this is tolerable. With `-bed_def none` I get :

```
Comparing beddef_lc_60km_out9.nc and results/beddef_lc_60km_outf.nc
Files are the same within tolerance 0.0e+00
```

**Comment by sebhinck** [comment]
Wow, that is great! I did not expect a perfect match! :)

To conclude,
- the start-year issue is due to rounding. By making sure that only whole seconds are hit this is fixed.
- the discrepancy between "one-go vs restart" is caused by the adaptive timestepping that might be different for for the two cases due to the restarts. By making sure that the runs stay in sync (i.e. `-time_stepping.hit_multiples`) the runs are identical (when using no beddef model)


One way to overcome this could be to always force to hit a multiple of one year (`-time_stepping.hit_multiples 1`). By doing this the maximum time step is limited to one year, but we make sure that all runs do exactly the same time steps (if the PISM run lengths are whole years and extra vars are written at the same dates).

**Comment by talbrecht** [comment]
@ckhroulev : Did these insights regarding reproducibility/restartability somehow merge into the dev? What would be the necessary steps?

**Comment by ckhroulev** [comment]
One of the issues described here should be resolved by #486. I seem to remember that one of you (@talbrecht, @tkleiner, or @sebhinck) brought up another one in this discussion as well. Would you mind pointing it out to me? Thanks!

**Comment by sebhinck** [comment]
Great that you implemented the more general case!

As far as I remember the other issue was related to creation of restart files. If you write restart files during the run, these are by default written close to the requested time, but do not force the timestep to be adapted.
If you, however, end the run at a certain time and restart the model from this restart file, the model hits exactly the requested time and thus both restart files (save_file & out_file) are slightly off.
If you want to compare such experiments, you should make sure to force hitting the requested save time (`save_force_output_times`).

It should also be noted that requesting writing of  "extra" files impacts the timestepping and that for reproducability reasons, you have to make sure that your runs write these files at the same time steps (or you disable `extra_force_output_times`).

**Comment by ckhroulev** [comment]
@sebhinck To me these sound like topics that should be explained in the manual (specifically, the fact that one needs to be careful when comparing results from stopped-and-restarted simulations to ones that ran without stopping), but I can't think of anything I need to fix in the code.

[This section](https://pism.github.io/pism/manual/practical-usage/time-stepping.html) may be a good place for this. As always, contributions are welcome.

**Comment by ckhroulev** [comment]
As far as I can tell this issue is fixed by #486.

Please create a separate issue on GitHub if you think that there's something else in the discussion above that we need to address.
